File: next-permutation/next-permutation.py
def next_permutation(arr):
    for k in xrange(len(arr)-2, -1, -1):
        if arr[k]<arr[k+1]:
            i = len(arr)-1
            while arr[i]<=arr[k]:
                i -= 1
            arr[k], arr[i] = arr[i], arr[k]
            arr[k+1:] = arr[k+1:][::-1]
            return True
    return False

def main():
    arr = [1, 2, 2, 3, 4]
    while True:
        print arr
        if not next_permutation(arr):
            break

if __name__ == '__main__':
    main()

File: next-permutation/next-permutation.cpp
#include "bits/stdc++.h"

int main(){
    std::vector<int> arr;
    arr.push_back(1);
    arr.push_back(2);
    arr.push_back(2);
    arr.push_back(3);
    arr.push_back(4);
    do{
        for(const int x : arr){
            printf("%d ", x);
        }
        printf("\n");
    }while (std::next_permutation(arr.begin(), arr.end()));
    return 0;
}
